artifact/tar/platform/libarchive/tar.cpp

```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <cstdint>
+#include <vector>
+#include <algorithm>
+#include <string>
+#include <iostream>
+
+#include <common/io.hpp>
+#include <common/log.hpp>
+#include <artifact/tar/tar.hpp>
+
+namespace mender {
+namespace tar {
+
+using namespace std;
+
+namespace log = mender::common::log;
+
+const ErrorCategoryClass ErrorCategory {};
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "TarErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	case TarReaderError:
+		return "Error reading the tar stream";
+	case TarShortReadError:
+		return "Short read error";
+	case TarEntryError:
+		return "Error reading the tar entry";
+	case TarEOFError:
+		return "Archive EOF reached";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code) {
+	return error::Error(error_condition(code, ErrorCategory), "");
+}
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+mender::common::expected::Expected<size_t, mender::common::error::Error> Entry::Read(
+	vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) {
+	expected::ExpectedSize read_bytes = reader_.Read(start, end);
+
+	if (!read_bytes) {
+		return read_bytes;
+	}
+
+	if (read_bytes.value() < 0) {
```

Comment:
`ExpectedSize` uses `size_t`, so it can't be below zero. Surprised it didn't warn about this..